src/Core/PolynomialFunction.java

```diff
@@ -129,31 +132,98 @@ public void derive() {
+            if (Math.abs(y) < EPSILON) {
+                double root = formatDouble(x);
+                if (shouldAddValue(root, roots)) {
+                    roots.add(root);
+                }
+            // First use Incremental Search to check for a sign change (if-statement)
+            // then calculate an approximately really close position using linear interpolation
+            } else if (prevY * y < 0) {
+                double root = (x - step) + (x - (x - step)) * Math.abs(prevY) / (Math.abs(prevY) + Math.abs(y));
+                root = formatDouble(root);
+                if (shouldAddValue(root, roots)) {
+                    roots.add(root);
+                }
             }
-            prevPoint = point;
+            prevY = y;
         }
+        return roots;
     }
 
-    public void calcExtremePoints(double start, double end, double step) {
-        extremePoints = new ArrayList<>();
-        Vector2D startPoint = evaluate(start);
-        Vector2D nextPoint = evaluate(start + step);
-        double prevSlope = (nextPoint.y - startPoint.y) / step;
-        for (double x = start + 2 * step; x <= end; x += step) {
-            Vector2D point = evaluate(x);
-            double slope = (point.y - evaluate(x - step).y) / step;
-            if (prevSlope * slope <= 0) {
-                extremePoints.add(x - step);
+    private Set<Double> calcExtremes(double start, double end, double step) {
+        Set<Double> extremes = new LinkedHashSet<>();
+        double prevSlope = evaluateDerivative(start);
+        for (double x = start + step; x <= end; x += step) {
+            double slope = evaluateDerivative(x);
+            // If the value of the derivative is incredibly close to zero, it is an extreme point => Add
+            if (Math.abs(slope) < EPSILON) {
+                double extreme = formatDouble(x);
+                if (shouldAddValue(extreme, extremes)) {
+                    extremes.add(extreme);
+                }
+            // First use Incremental Search to check for a sign change (if-statement)
+            // then calculate an approximately really close position using linear interpolation
+            } else if (prevSlope * slope < 0) {
+                double extreme = (x - step) + (x - (x - step)) * Math.abs(prevSlope) / (Math.abs(prevSlope) + Math.abs(slope));
+                extreme = formatDouble(extreme);
+                if (shouldAddValue(extreme, extremes)) {
+                    extremes.add(extreme);
+                }
             }
             prevSlope = slope;
         }
+        return extremes;
+    }
+
+    private boolean shouldAddValue(double value, Set<Double> existingValues) {
+        for (Double existingValue : existingValues) {
+            if (Math.abs(existingValue - value) < EPSILON) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private double formatDouble(double value) {
```

Comment:
the name of this function does not clearly convey what it does. If I understand correctly, it more or less 'squishes' the number into 6 decimals after the comma? If yes, it should be renamed to include this 'rounding' property. Also, since our EPSILON is *really* tiny, and we are rounding to a fraction magnitudes larger, the unique value property of the Set<Double> makes checking 'shouldAddValue' redundant. Just add it to the set, and it will reject it if it's too close (within 6 decimals) to any other contained number. Then, we can remove 'shouldAddValue' completely.

Reply:
Good thing you pointed that out with shouldAddValue because that was also a relict. I had trouble even with the set with reoccurring values. That was before epsilon was integrated. 

As for formatDouble it was literally at the beginning only a function to make the  -0.00 to 0.00. And got later used to correct weird number "glitches". Will adapt the name to roundDouble. Thanks for the catch. Done